coding.md

```diff
@@ -0,0 +1,109 @@
+
+* Boolean variables should always be ran through the `|bool` filter to
+  normalize the input into a Boolean.
+
+* Do not rely on variable settings in an example env defaults file. These
+  files are just for example and may not be used in production. Instead rely
+  on role defaults or the `|default()` filter to set a default value.
+
+* Filter application should be done without spaces between the variable,
+  the pipe symbol, and the filter name: `{{ foo|default(False)|bool }}`
+
+* Be aware of jinja2 python object method capabilities when deciding on
+  variable names. Avoid any variable name that could be confused with an
+  object method (like endswith, startswith, split, upper, index, count).
+
+## Tasks
+* When possible, use full YAML syntax (`key: value`) for task arguments,
+  instead of `key=value` pairs. Some task inputs will require full yaml,
+  it is better to be consistent across the playbooks.
+
+* When writing a task that notifies handlers, the notify statement should be
+  last. Any tags above the notify. Any loop statement should come just before
+  the tags, and any conditional should go above that.
+
+* Task names must not use variables in them.
+
+* All Tasks, Plays, and Handlers must have a name.
+
+* Do not use `ignore_errors`, instead use `failed_when: false`. This prevents
+  red output from being shown on screen, which makes debugging difficult.
+
+* Tasks which just gather data and do not cause change should not register
+  "change" from ansible point of view. Make use of `changed_when: False`.
+
+* No tasks should prompt the user for interaction. The intention of Ursula
+  is to run completely non-interactive.
+
+* Tag expression should be either a single word `tags: thistag`, or a YAML
+  list of tags to apply with each tag on a new line. Do not use `tags=foo`,
+  particularly when including a task file.
+
+* Do not rely on a when conditional to identify that a loop variable is
+  undefined to avoid the task. Use a filter to default the variable to an
+  empty list: `with_items: "{{ foo|default([]) }}"`
+
+## Roles
+* Role dependencies have a cost. Be wary of introducing a new role
+  dependency, particularly if it could be ran from many plays, or if the
+  dependency role includes a large number of tasks.
+
+* Task file includes from a role's `tasks/main.yml` should be avoided
+  unless there is a desire to apply a tag or a conditional to an entire
+  set of logically grouped tasks.
+
+## Modules
+* Prefer the command module over the shell module.
+
+* Files are put in place via the template module, rather than copy, so that
+  all the input files that go to a system are gathered in the same directory
+  (`templates/`)
+
+* Directory structure within templates/ must match (as close as possible) the
+  destination path of the file in question. e.g. if the destination is
+  `/etc/nova/nova.conf`, the file would live in `templates/etc/nova/nova.conf`
+  (note the lack of `.j2` suffix)
+
```

Comment:
We should have guidelines here about modules we write, too. For example, what pep8 codes we ignore, if any. I'd suggest running pylint too.